tiledb/bioimg/visualization/helpers.py

```diff
@@ -0,0 +1,65 @@
+import json
+from dataclasses import dataclass
+from typing import List, Optional, Sequence
+
+import tiledb
+
+from ..converters.axes import Axes
+
+
+@dataclass(frozen=True)
+class ZoomLevelRecord:
+    zoom_level: int
+    image_level: int
+    downsample: int
+    width: int
+    height: int
+    namespace: str
+    array_uri: str
+    axes: str
+
+
+def calculate_metadata(
+    namespace: str, group_uri: str, config: Optional[tiledb.Config] = None
+) -> Sequence[ZoomLevelRecord]:
+    """
+    Calculates the required metadata for the Tile Image viewer.
+
+    :param namespace: the namespace of the group
+    :param group_uri: path to the group
+    :param config: the configuration for constructing the tileDB context
+    """
+    metadata: List[ZoomLevelRecord] = []
+
+    context = None if config is None else tiledb.Ctx(config)
+    group = tiledb.Group(group_uri, "r", ctx=context)
+    zoom_level = 0
+
+    for meta in json.loads(group.meta["levels"])[::-1]:
+        axes = Axes(meta["axes"])
+        axes_mapper = axes.mapper(Axes("XYC"))
+        shape = axes_mapper.map_shape(meta["shape"])
+
+        zoom_factor = round(shape[0] / metadata[-1].width) if len(metadata) > 0 else 2
+
+        if not ((zoom_factor & (zoom_factor - 1) == 0) and zoom_factor != 0):
```

Comment:
Every time (thankfully rarely) I see a bitwise `&` or `|` I have to spend several seconds/minutes to figure out:
- (a) if this was really meant to be logical `and` or `or`
- (b) or if it is intentionally bitwise operator, and if yes why
- (c) if it [makes a difference](https://stackoverflow.com/questions/3845018/boolean-operators-vs-bitwise-operators) in this case or not:
```py
In [1]: 4 & 0, 4 and 0
Out[1]: (0, 0)

In [2]: 4 & 1, 4 and 1
Out[2]: (0, 1)
```

TL;DR: Do **not** use bitwise `and/or` without a very good reason, and if there is a very good reason (e.g. bit fiddling) please add a comment or two.